src/backoff.rs

```diff
@@ -0,0 +1,242 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+// https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+
+use rand::{thread_rng, Rng};
+
+pub trait Backoff: Clone + Copy + Send + Sync + 'static {
+    fn next_delay_ms(&mut self) -> Option<u64>;
+
+    fn exponential_delay_ms(base_delay_ms: u64, max_delay_ms: u64, attempts: u32) -> u64 {
+        max_delay_ms.min(2u64.pow(attempts) * base_delay_ms)
+    }
+}
+
+#[derive(Clone, Copy)]
+pub struct BackoffState {
+    current_attempts: u32,
+    max_attempts: u32,
+    base_delay_ms: u64,
+    max_delay_ms: u64,
+    last_delay_ms: u64,
+}
+
+impl BackoffState {
+    pub fn new(base_delay_ms: u64, max_delay_ms: u64, attempts: u32) -> Self {
+        // Both argument `base_delay_ms` and `max_delay_ms` cannot be less than
+        // 2 to prevent panic on `rng.gen_range`.
+
+        let base_delay_ms = match base_delay_ms {
+            0 | 1 => 2,
+            v => v,
+        };
+
+        let max_delay_ms = match max_delay_ms {
+            0 | 1 => 2,
+            v => v,
+        };
```

Comment:
These checks only apply on jitter backoffs? Maybe we can check them only when creating those kinds of backoffs.

Reply:
I think it's all right to just panic on bad arguments. Implicit fixes may bring unexpected result.